drivers: spi: gd32: Add support DMA transfer

soburi opened the pull request:
Add supporting DMA-based transfer for GD32 SPI.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

mbolivar-nordic reviewed (commented):
@teburd -- I'm trying to take a step back from reviewing subsystem-specific DT changes in favor of asking those subsystem maintainers to take this on. I'm going to leave this up to you -- please feel free to ping if you have any specific questions

nandojve commented:
Hi @soburi ,

You can rebase since #47501 got merge.

soburi commented:
Hi, @nandojve , @cameled
Thank you for your continued support.
As with #47501, please review it.

galak commented on dts/bindings/spi/gd,gd32-spi.yaml:
```diff
@@ -24,3 +24,23 @@ properties:
   pinctrl-names:
     required: true
+
+  tx-dma-priority:
```
Why isn't this using a `dmas` property like dts/bindings/spi/atmel,sam0-spi.yaml

soburi replied:
I will apply your pointing out.
dma-gd32.c need fix prior than spi,
so I create new PR #48756.
Could you please review also it?

galak reviewed (dismissed):
Use standard `dmas` property.

cameled commented:
@soburi In general, gd32 evaluate boards have a spi flash. I have tried to run dma on that spi flash from gd32f450i_eval board, but looks there have something blocked.  I will try to dig out tomorrow. Anyway, this is an impressive work.

cameled reviewed (commented):
@soburi , Please add below patch to support run `samples/drivers/spi_flash` sample with DMA on `gd32f450i_eval` board.

```
diff --git a/boards/arm/gd32f450i_eval/gd32f450i_eval.dts b/boards/arm/gd32f450i_eval/gd32f450i_eval.dts
index cf90ba9b39..2ca2443c87 100644
--- a/boards/arm/gd32f450i_eval/gd32f450i_eval.dts
+++ b/boards/arm/gd32f450i_eval/gd32f450i_eval.dts
@@ -144,11 +144,19 @@
 	};
 };
 
+&dma1 {
+	status = "okay";
+};
+
 &spi5 {
 	status = "okay";
 	pinctrl-0 = <&spi5_default>;
 	pinctrl-names = "default";
 	cs-gpios = <&gpioi 8 GPIO_ACTIVE_LOW>;
+	dmas = <&dma1 5>, <&dma1 6>;
+	dma-names = "tx", "rx";
+	tx-dma-peripheral = <1>;
+	rx-dma-peripheral = <1>;
 
 	nor_flash: gd25q16@0 {
 		compatible ="jedec,spi-nor";
diff --git a/samples/drivers/spi_flash/boards/gd32f450i_eval.conf b/samples/drivers/spi_flash/boards/gd32f450i_eval.conf
index b5aa8903ef..6d3d1db11f 100644
--- a/samples/drivers/spi_flash/boards/gd32f450i_eval.conf
+++ b/samples/drivers/spi_flash/boards/gd32f450i_eval.conf
@@ -5,3 +5,4 @@
 #
 
 CONFIG_SPI_NOR=y
+CONFIG_SPI_GD32_DMA=y
```

cameled commented on dts/bindings/spi/gd,gd32-spi.yaml:
```diff
@@ -24,3 +24,23 @@ properties:
+    description: Priority of DMA for receive
+    default: 0
+
+  tx-dma-peripheral:
```
Please refer https://github.com/zephyrproject-rtos/zephyr/pull/42215#discussion_r827790806. 

We need to provide a simple way for the user to select the peripheral easily.

cameled replied:
This can be done in another PR, not related to this.

cameled commented:
Please consider to split the original commit with `drivers` and `dts`.

nandojve commented:
Hi @soburi ,

Do you think can finish in the next 2 days? Sep. 9th is last day to get it in otherwise only for next release.

soburi commented:
Hi, @nandojve 

> Hi @soburi ,
> 
> Do you think can finish in the next 2 days? Sep. 9th is last day to get it in otherwise only for next release.

I can rewrite the source with in 2 days, but #48756 discussion may not solve.
I will prepare, but I think it will be difficult for the above reasons.

soburi commented:
Mark as draft until #48756 solved.

cameled commented on dts/bindings/dma/gd,gd32-dma.yaml:
```diff
@@ -4,6 +4,79 @@
+  The first cell assign channel 3 to receive, channel 5 assigned to transmit.
+  The second cell is slot. Both channels select 3.
+  Slot is effective only for GD32F4XX series. ignored in otherwise.
+  What the slot number '3' meaning depends on the DMA controller and channel.
```
From GD32F4xx User Manual, slot (PERIEN btis) can be determined by Channel and Peripheral. For example, DMA0 Channel 0 and SPI2_RX only have one PERIEN, which value is 0.

After that, we can define some helper macro, like DMA0_CH0_SPI2_RX, user can easily to select them.  helper macro implement can refer `include/zephyr/dt-bindings/reset/gd32f3x0.h`.

soburi replied:
Already solved in #48756.

cameled commented on dts/bindings/dma/gd,gd32-dma.yaml:
```diff
@@ -4,6 +4,79 @@
+                 - 0x2: PERIPH to MEMORY
+                 - 0x3: reserved for PERIPH to PERIPH
+
+        - bit 9: Peripheral address increase
```
Please fix the alignment.

soburi replied:
Already solved in #48756.

cameled commented on boards/arm/gd32f450i_eval/gd32f450i_eval.dts:
```diff
@@ -143,11 +143,19 @@
 	pinctrl-0 = <&spi5_default>;
 	pinctrl-names = "default";
 	cs-gpios = <&gpioi 8 GPIO_ACTIVE_LOW>;
+	dmas = <&dma1 5>, <&dma1 6>;
```
Patch to sync with new dma cells setting.
```
diff --git a/boards/arm/gd32f450i_eval/gd32f450i_eval.dts b/boards/arm/gd32f450i_eval/gd32f450i_eval.dts
index ba87bb8e1f..a9614525d2 100644
--- a/boards/arm/gd32f450i_eval/gd32f450i_eval.dts
+++ b/boards/arm/gd32f450i_eval/gd32f450i_eval.dts
@@ -152,10 +152,8 @@
 	pinctrl-0 = <&spi5_default>;
 	pinctrl-names = "default";
 	cs-gpios = <&gpioi 8 GPIO_ACTIVE_LOW>;
-	dmas = <&dma1 5>, <&dma1 6>;
+	dmas = <&dma1 5 1 0 0>, <&dma1 6 1 0 0>;
 	dma-names = "tx", "rx";
-	tx-dma-peripheral = <1>;
-	rx-dma-peripheral = <1>;
 
 	nor_flash: gd25q16@0 {
 		compatible ="jedec,spi-nor";
```

nandojve commented:
> Hi, @nandojve
> 
> > Hi @soburi ,
> > Do you think can finish in the next 2 days? Sep. 9th is last day to get it in otherwise only for next release.
> 
> I can rewrite the source with in 2 days, but #48756 discussion may not solve. I will prepare, but I think it will be difficult for the above reasons.

Hi @soburi ,

I think it is not necessary to be in rush. We already made very good progress for this release.

soburi commented:
> This pull request has been marked as stale because it has been open (more than) 60 days with no activity. Remove the stale label or add a comment saying that you would like to have the label removed otherwise this pull request will automatically be closed in 14 days. Note, that you can always re-open a closed pull request at any time.

Rebase to solve the stall.
It is still waiting  #48756 discussion result.

nandojve commented:
> > This pull request has been marked as stale because it has been open (more than) 60 days with no activity. Remove the stale label or add a comment saying that you would like to have the label removed otherwise this pull request will automatically be closed in 14 days. Note, that you can always re-open a closed pull request at any time.
> 
> Rebase to solve the stall. It is still waiting #48756 discussion result.

Hi @soburi,
In this case, ask feedback for @person to clarify missing points.
Remember that DMA is quite complex drivers and it must have tests to cover it. SPI have their own tests with DMA  at `tests/drivers/spi/spi_loopback`.

soburi committed:
boards: gd32f450i_eval: Enable DMA transfer for nor-flash

Enable DMA transfer via SPI bus for nor-flash.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
drivers: spi: gd32: Add support DMA transfer

Add supporting DMA-based transfer for GD32 SPI.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
tests: spi: spi_loopback: add interrupt and DMA tests for GD32 boards

Add interrupt and DMA tests for GD32 boards.
Use overlay-gd32-spi-interrupt.conf when testing interrupt based transfer,
use overlay-gd32-spi-dma.conf when testing DMA transfer,

Add overlay file for GD32 boards for these tests.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
boards: gd32: add `spi` as supported feature

Add `spi` as supported feature to yaml configuration.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

cameled commented on drivers/spi/spi_gd32.c:
```diff
@@ -195,6 +233,146 @@ static int spi_gd32_frame_exchange(const struct device *dev)
+	struct dma_status stat;
+	int ret = 0;
+
+	if (!cfg->dma[TX].dev && spi_context_tx_on(&data->ctx)) {
```
Can we move dma device check out of this function?

From below code, spi dma `TX` and `RX` enabled simultaneously.
```
	SPI_CTL1(cfg->reg) |= (SPI_CTL1_DMATEN | SPI_CTL1_DMAREN);
```

soburi replied:
DMA has been changed to guarantee its existence at initialization, so the check here is no longer necessary.
Removed from here.

cameled reviewed (changes requested):
Great work, some need to confirm.

soburi commented on drivers/spi/spi_gd32.c:
```diff
@@ -219,31 +397,51 @@ static int spi_gd32_transceive_impl(const struct device *dev,
-#else
-	do {
-		ret = spi_gd32_frame_exchange(dev);
+	if (spi_gd32_dma_enabled(dev)) {
```
This check is necessary because DMA transfers may not be enabled for all SPIs even if DMA configuration enabled.

cameled replied:
Suggestion:
```
#ifdef CONFIG_SPI_GD32_INTERRUPT
	if (IS_ENABLED(CONFIG_SPI_GD32_DMA) && spi_gd32_dma_enabled(dev)) {
		for (size_t i = 0; i < ARRAY_SIZE(data->dma); i++) {
			data->dma[i].count = 0;
		}

		ret = spi_gd32_start_dma_transceive(dev);
		if (ret < 0) {
			goto dma_error;
		}
		ret = spi_context_wait_for_completion(&data->ctx);
	} else {

		SPI_STAT(cfg->reg) &=
			~(SPI_STAT_RBNE | SPI_STAT_TBE | SPI_GD32_ERR_MASK);
		SPI_CTL1(cfg->reg) |=
			(SPI_CTL1_RBNEIE | SPI_CTL1_TBEIE | SPI_CTL1_ERRIE);
		ret = spi_context_wait_for_completion(&data->ctx);
	}
#else

	do {
		ret = spi_gd32_frame_exchange(dev);
		if (ret < 0) {
			break;
		}
	} while (spi_gd32_transfer_ongoing(data));

#ifdef CONFIG_SPI_ASYNC
	spi_context_complete(&data->ctx, dev, ret);
#endif
#endif
```

soburi replied:
Applied it with some modifications to solve compiler warnings.

cameled commented on drivers/spi/spi_gd32.c:
```diff
@@ -195,6 +233,146 @@ static int spi_gd32_frame_exchange(const struct device *dev)
+#ifdef CONFIG_SPI_GD32_DMA
+	const struct spi_gd32_config *cfg = dev->config;
+
+	if (cfg->dma[TX].dev || cfg->dma[RX].dev) {
```
Shall we use `&&` in here? TX and RX dma can not work individually.

soburi replied:
Change to require setting both TX and RX DMA channels.
Change to && here.

soburi commented on drivers/spi/spi_gd32.c:
```diff
@@ -304,6 +515,86 @@ static void spi_gd32_isr(struct device *dev)
+	struct spi_gd32_dma_data *dma = data->dma;
+	const size_t chunk_len = spi_context_max_continuous_chunk(&data->ctx);
+
+	if (chunk_len <= 0) {
```
Removed this check.

soburi commented on drivers/spi/spi_gd32.c:
```diff
@@ -37,18 +41,52 @@ LOG_MODULE_REGISTER(spi_gd32);
+#ifdef CONFIG_SPI_GD32_DMA
+
+enum spi_gd32_dma_direction {
+	RX,
```
Applied it.

soburi commented on drivers/spi/spi_gd32.c:
```diff
@@ -195,6 +233,146 @@ static int spi_gd32_frame_exchange(const struct device *dev)
+	struct dma_block_config *block_cfg = &data->dma[dir].block;
+	const struct spi_gd32_dma_config *dma = &cfg->dma[dir];
+	static uint32_t dummy_tx;
+	static uint32_t dummy_rx;
```
Change to declare as a global static variable.

soburi replied:
Applied it with some modifications to solve compiler warnings.

cameled commented on drivers/spi/spi_gd32.c:
```diff
@@ -37,20 +41,74 @@ LOG_MODULE_REGISTER(spi_gd32);
 
+#ifdef CONFIG_SPI_GD32_DMA
+
+static uint32_t dummy_tx;
+static uint32_t dummy_rx;
```
We should not access these from MCU, because DMA may access it at any time after spi driver initialized. 

SRAM can only be accessed by one AHB master at a time. So `dummy_tx` and `dummy_rx` can merge as one.

soburi replied:
@nandojve 
Static variable initialize default with zero. So I think it is no problem.
https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf §6.7.8/10
(It is rare to initialize a static variable explicitly with zero in zephyr.)

@cameled 
I want to fix the rx_dummy to zero.
Transmitting modifies the value if we merge tx_dummy and rx_dummy into a single variable.
I think it may make it harder to debug.

cameled replied:
Hi @soburi , I update the above comment, as @nandojve suggested, we can give an explicit default zero to `dummy_tx`.

And you are right, `dummy_tx` and `dummy_rx` can not merge as one.

nandojve replied:
Fine for me, I just rose to make sure you folks prefer 0x00 or 0xFF.

cameled commented on drivers/spi/spi_gd32.c:
```diff
@@ -195,6 +253,115 @@ static int spi_gd32_frame_exchange(const struct device *dev)
+{
+	const struct spi_gd32_config *cfg = dev->config;
+	struct spi_gd32_data *data = dev->data;
+	const uint32_t datareg = (uint32_t)&SPI_DATA(cfg->reg);
```
Suggestion: Remove `datareg`, asign `SPI_DATA` address directly.

soburi replied:
Applied your suggestion, Thank you.

cameled commented on drivers/spi/spi_gd32.c:
```diff
@@ -195,6 +253,115 @@ static int spi_gd32_frame_exchange(const struct device *dev)
+static void spi_gd32_dma_callback(const struct device *dmadev, void *arg,
+				  uint32_t channel, int status);
+
+static uint32_t spi_gd32_dma_setup(const struct device *dev, uint32_t dir)
```
Suggestion: Give `dir` a const, prevent unexpected change.

soburi replied:
Applied your suggestion, Thank you.

cameled commented on drivers/spi/spi_gd32.c:
```diff
@@ -195,6 +253,115 @@ static int spi_gd32_frame_exchange(const struct device *dev)
 }
 
+#ifdef CONFIG_SPI_GD32_DMA
+static void spi_gd32_dma_callback(const struct device *dmadev, void *arg,
```
Suggestion: change `dmadev` to `dma_dev`,

soburi replied:
Applied your suggestion, Thank you.

cameled reviewed (commented):
LGTM, add some suggestion.

cameled commented on drivers/spi/spi_gd32.c:
```diff
@@ -37,20 +41,74 @@ LOG_MODULE_REGISTER(spi_gd32);
+	return false;
+}
+
+static size_t spi_gd32_dma_enabled_num(const struct device *dev)
```
Compared with `ARRAY_SIZE`, function call have an extra overhead.

soburi replied:
Determining from ARRAY_SIZE won't work if both exists DMA enabled SPI and disabled one.

nandojve reviewed (approved):
Thank you @soburi !
